### PR TITLE
Support fusion of mutable operators

### PR DESCRIPTION
**Pull request by apaszke**
This is a next step in making fusions in ResNets possible. The only piece
we're missing apart from my open PRs is the ability to differentiate convolutions, which I'm hoping to implement soon (using a method similar to the one I used for BN).

Note that while the method is valid, the fusions usually will only
happen in traced graphs. That is because while tracing, no matter
how the expression looks like, we will always assign the `Value*` of
the mutable _output_ to the tensor. On the other hand, the code
usually contains expressions like `x.relu_()`, which when scripted
does not overwrite the value of `x` with the return from that call
(which is exactly the same at run-time, but a different `Value*`).
This means that in this case there are no data dependencies that
our fuser could follow, and it will never even consider such a node
for fusion, even though it would result in a nice group.

**tl;dr** The following function is currently perfectly fusable when
traced, but not when scripted:

```python
def f(x):
    y = x * 2
    y.relu_()
    return y * 2
```

cc @suo @zdevito @mruberry

### CODE REVIEW

File: torch/csrc/jit/passes/graph_fuser.cpp [hunk 1]
```diff
@@ -115,6 +115,119 @@ bool isSimpleMap(Node *node) {
...
+  node->output()->replaceAllUsesWith(output);
+  node->destroy();
+  return output;
+}
+
+bool needsToMutate(Node *node, const AliasDb& aliasDb) {
+  // A node is safe to be de-inplaced if all uses of values
+  // that can alias its output (that are not its output) happen
+  // before this node.
+  Value * output = node->output();
```
Comment (suo) [inline torch/csrc/jit/passes/graph_fuser.cpp]: You'll also need to check that there is no use of a wildcard value after `node`. I have a PR soon to hide "wildcards" as an implementation detail to avoid making people remember this in the future 😛

Comment (apaszke) [reply torch/csrc/jit/passes/graph_fuser.cpp]: I'd really like us to relax the definition of a wildcard to simply "all outputs alias all inputs and outputs", because as it is right now is super weird. It's not like we'll be manufacturing tensors you don't even get as arguments out of thin air...

Anyway, how do I get a list of wildcard values?

Comment (suo) [reply torch/csrc/jit/passes/graph_fuser.cpp]: The main place that wildcards show up is because we don't yet track aliases "into" lists and tuples. So when we index into a list and retrieve a tensor, it's not enough to say that it may alias all the inputs/outputs of the node.

I think it's enough to getWildcardNodes() and check that they are all before `node`.

**Comment by apaszke** [comment]
I've added checks for mutable nodes, but unfortunately the interaction of the fuser with alias db is pretty terrible. The `isBefore` checks often run into the problem that some of the nodes are deleted from the main graph, but their pointers stay in alias database maps, so we end up accessing freed up (and possibly reused) memory... We need a way to keep the DB up to date after all the moves we do.

File: torch/csrc/jit/passes/graph_fuser.cpp [hunk 2]
```diff
@@ -128,6 +223,7 @@ Value* broadcastSizes(at::ArrayRef<Value*> sizes) {
 struct GraphFuser {
   Block* block_;
   std::shared_ptr<Graph> graph_;
+  c10::optional<AliasDb> aliasDb_;
```
Comment (suo) [inline torch/csrc/jit/passes/graph_fuser.cpp]: Doesn't need to be optional, we can just initialize it in the constructor

**Comment by suo** [comment]
Yeah I'm working on that next for mutability.

**Review by suo — DISMISSED** [review]

**Comment by apaszke** [comment]
Can you please elaborate? Is "that" the ability to update the alias DB faster than O(n) as you modify the graph?

**Comment by suo** [comment]
Yep! I think it will probably end up looking like: give AliasDb a set of methods to manipulate the graph. As long as you only mutate the graph through those methods, you can be sure that everything is "mutability-safe".

**Comment by apaszke** [comment]
Well the CI failures on this thing are because of accesses to freed memory via AliasDb pointers, so I don't think we can do anything else than wait for your patch.

**Review by ezyang — CHANGES_REQUESTED** [review]
needs to be rebased over the alias patch
